### PR TITLE
Upgrade collaboration service's existing functionality

### LINKED ISSUE #125: Implement a form of conflict resolution for collaboration service
Currently, when two users are connected and typing at the same time, the update will cause the other user's cursor to jump to end of the code editor. 

- Implement a form of conflict resolution that will only update based on the delta of the content on the editor (ensuring the user's cursor remain where they are typing)
- If possible, track the other user's cursor as well when he is on the editor panel

### LINKED ISSUE #128: Bugs found in collaboration service
## Bug 1:
1. Matched two users.
2. For one user, navigate to the dashboard page.
3. Go back to the room page again, you will get 500, error indicating that:
```
Too many re-renders. React limits the number of renders to prevent an infinite loop.
```

This is caused by 2 components: `CodeEditor` and `ChatSpace`, and they are related to socket events update and React state update (the dispatch function). There might be logic flaw or wrong implementation, need someone to take a deeper look and fix the issue. 

## Bug 2:

Ctrl A + Del/Backspace will cause the editor to crash/unmount

### PR DESCRIPTION
**Pull request by weidak**
# Pull Request

## Description

This PR improves the functionality provided by collaboration service and fixes some bugs discussed previously.

## Related Issue(s)

Closes #128 
Closes #125 

## Changes Made

Previously, I was using useState to set the content of the editor and it was causing a cursors to jump and texts to flash white (since it was constantly resetting the editor's value).

With these new changes, I have implemented changes based on a specific range and offset, where in theory the editor is now able to update in isolation to the other user. We will have to test this out once this is merged to master and deployed to AWS for further investigation when latency is introduced. Also, I have added live cursors.

Fixed the bugs mentioned in #128 by removing `setError()` useState to fix build check, as we are now reliant on the `isNotFoundError()` that exists within collab context.

## Screenshots (if applicable)

Introducing live cursor and selections:

image

Shows the name of the other user on hover:

image

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

## Additional Notes/References

- To be tested with two pcs when merged to master and deployed to aws
- Add handler tests to integration test when after testing its functionality in the cloud so that it does not require more changes

### CODE REVIEW

**Review by c0j0s — APPROVED** [review]
LGTM

Minor, possible to stick the icons to one side? seem to be out of place when stretched.
image

File: frontend/src/components/collab/CodeEditor.tsx [hunk 1]
```diff
@@ -1,20 +1,169 @@
...
+    if (!editorRef.current) return;
+
+    decorations.current = editorRef.current.createDecorationsCollection([]);
+
+    let selectedRange: Range | null = null;
+
+    editor.onDidPaste((event: monaco.editor.IPasteEvent) => {
+      console.log(editor)
+      console.log("On did paste", event);
+    })
```
[the remark this answers is not in the document]
Comment (weidak) [reply frontend/src/components/collab/CodeEditor.tsx:120]: oops, this currently servers no purpose, forgot to remove. i was trying to solve a bug where when u drag the text around, the events are too quick to update both the deletion + insertion so it just performs insertion.. i still cant solve it though but its smth q small for now

**Comment by weidak** [comment]
> LGTM
> 
> Minor, possible to stick the icons to one side? seem to be out of place when stretched. image

sure, did not notice this when developing since it was always half half.. will update this to make the ui better

**Comment by weidak** [comment]
Updated UI:

image

**Pull request merged**